subsys/zbus/zbus.c

```diff
@@ -4,92 +4,98 @@
  */
 
 #include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/iterable_sections.h>
 #include <zephyr/logging/log.h>
 #include <zephyr/sys/printk.h>
 #include <zephyr/zbus/zbus.h>
 LOG_MODULE_REGISTER(zbus, CONFIG_ZBUS_LOG_LEVEL);
 
-#if (CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE > 0)
-static inline void _zbus_notify_runtime_listeners(const struct zbus_channel *chan)
+int _zbus_init(void)
 {
-	__ASSERT(chan != NULL, "chan is required");
-
-	struct zbus_observer_node *obs_nd, *tmp;
-
-	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(chan->runtime_observers, obs_nd, tmp, node) {
-
-		__ASSERT(obs_nd != NULL, "observer node is NULL");
-
-		if (obs_nd->obs->enabled && (obs_nd->obs->callback != NULL)) {
-			obs_nd->obs->callback(chan);
+	const struct zbus_channel *curr = NULL;
+	const struct zbus_channel *prev = NULL;
+
+	STRUCT_SECTION_FOREACH(zbus_channel_observation, observation) {
+		curr = observation->chan;
+
+		if (prev != curr) {
+			if (prev == NULL) {
+				*curr->observers_start_idx = 0;
+				*curr->observers_end_idx = 0;
+			} else {
+				*curr->observers_start_idx = *prev->observers_end_idx;
+				*curr->observers_end_idx = *prev->observers_end_idx;
+			}
+			prev = curr;
 		}
+
+		++(*curr->observers_end_idx);
 	}
```

Comment:
Ok so this is needed to handle the channels with a single iterable section right? Had a really hard time understanding the whole chain of const pointer to const struct with const pointer to uint16_t variables - I suspect that this can be simplified a lot - but besides that, what's the ram/flash cost of maintaining all these structures?

The practical use case is this: we had a similar schema for the event system in our EC application, the implementation was way simpler but the concept was the same: we link the callback structures in the right order at system init and then we can traverse that efficiently in runtime. In the end, it turned out to waste too much RAM (we have hundreds of event callbacks so even 4 bytes each is significant) and we converged to a solution where every event has its own section (which was not aproblem to manage), the implementation became way simpler and much more efficient, both in term of memory and no need to do this dance at boot time.

I would suggest the same, the general feeling I have is that this is way too complicated (well, at least I have a hard time going through it) and has significant resource overhead (which admittedly I have a hard time quantifying).

I would recommend considering an approach where channels have their own section, making use of the existing `ITERABLE_SECTION_RAM_NUMERIC` functionality.

Reply:
I can try to improve that. However, I still think it is worth keeping complexity into the "dance at boot time" rather than letting the user deal with it. Again, maybe you, as an experienced developer, don't think LD files as a problem. But I think it is better to keep inexperienced developers away from that as possible. In my head, all the people would use ZBus. It makes the developer's life easier. Imagine a first-come dev trying to use ZBus and facing issues because he/she is not writing the LD file properly. It is a really hard issue to solve when you don't have experience. So, again, I prefer to keep complexity inside than outside. Regarding the RAM waste, this is really an issue. I will investigate alternatives to that.

Reply:
I don't think it's a problem to ask a user to follow some documentation to add a single line on a single file just because the file name happens to end in "ld" rather than "c".

I guess we should think of what tradeoff are taking with the project, is it an educational beginner friendly RTOS or "a scalable real-time operating system (RTOS) supporting multiple hardware architectures, optimized for resource constrained devices, and built with security in mind"?

Reply:
The solution I am proposing is as performant as the one you are suggesting. The only difference is the initial math, which is not complex. My solution is based on the ZMK approach, as you suggested before. I am only improving the way they do. During the publication, they traverse the whole observation array. In ZBus, we only traverse the observation list related to the channel instead, which is quicker. The current solution does not oppose the statement: "a scalable real-time operating system (RTOS) supporting multiple hardware architectures, optimized for resource-constrained devices, and built with security in mind". It only avoids the extra effort of creating a new file with duplicated information. In the way you are suggesting, the developer will need to define the channel in the C code and again declare that on the LD file. I cannot see a significant advantage in that. This may be a point we can discuss at the Architecture meeting.